netebpfext/net_ebpf_ext_helpers.h

```diff
@@ -0,0 +1,42 @@
+// Copyright (c) Microsoft Corporation
+// SPDX-License-Identifier: MIT
+
+/**
+ * @file Prototypes and implementations for helper functions that are common to various programs types implemented in
+ * netebpfext.
+ */
+
+#pragma once
+#include "ebpf_platform.h"
+
+static int
+_partial_sum(_In_reads_bytes_(buffer_length) const void* buffer, int buffer_length)
+{
+    int partial_sum = 0;
+    for (int i = 0; i < buffer_length / 2; i++)
+        partial_sum += *((uint16_t*)buffer + i);
+    return partial_sum;
+}
+
+static int
+_net_ebpf_ext_csum_diff(
+    _In_reads_bytes_opt_(from_size) const void* from,
+    int from_size,
+    _In_reads_bytes_opt_(to_size) const void* to,
+    int to_size,
+    int seed)
+{
+    int csum_diff = -1;
+
+    if ((from_size % 4 != 0) || (to_size % 4 != 0))
+        // size of buffers should be a multiple of 4.
+        goto Exit;
```

Comment:
How do you differentiate errors? The real csum diff could be -1.

Reply:
Thats a great point. I am stumped. From: https://www.man7.org/linux/man-pages/man7/bpf-helpers.7.html
`Return The checksum result, or a negative error code in case of failure.`
Suggestions?

Reply:
The checksum diff will never be likely a +ve number bigger than 1FFFF, so what if we return 0x2000?

Reply:
The real csum diff should always be a positive number (and indeed I think checksums are 16 bits so all you care about is the mod 16 result, so always between 0 and 0xffff).  Negative numbers are errors.

Reply:
The algorithm is seed - partial_sum(from) + partial_sum(to).

If seed = 0 and to == null, then?

Reply:
to == null would be an error from my reading of the man page, so one should not do any checksum computation if you pass null for either from or to

Reply:
>                   • With from_size > 0, to_size == 0 and seed set to
>                        checksum, it can be used when removing data from
>                        a packet.

So if seed = 0 and to_size == 0, then return seed - partial_sum(from) = (-partial_sum(from)) & 0xFFFF

Reply:
> The checksum diff will never be likely a +ve number bigger than 1FFFF, so what if we return 0x2000?

My comment above is wrong. It depends on the length of the buffer. Surely can be larger than 1FFFF - as this algorithm does not fold. Only the caller folds.

Reply:
This is an example of the code generating negative csum_diff

```
    uint16_t from_buffer[] = {0x4500, 0x0073, 0x0000, 0x4000, 0x4011, 0x0000, 0x2000, 0x0001, 0x2000, 0x000a};

    int csum2 = test_xdp_helper_t::csum_diff(from_buffer, sizeof(from_buffer), nullptr, 0, 0);
    printf("csum %d(0x%x)\n", csum2, csum2);

```
output: `csum -66959(0xfffefa71)`

Reply:
but does it ever return negative on Linux?  We should behave the same as whatever the answer is.

Reply:
Posted question in #574 -- I think the above example I gave is not a valid use case. In that case the function should assert and return -EINVAL if result csum_delta is < 0.
I will update the PR - pending answer to the above.

Reply:
I finally have the answer I think. The algorithm is correct:
`csum_diff = seed - from + to;`
But here the subtraction should be 1's complement and not 2's complement. In other words
`csum_diff = seed + (~from) + to;`

This should always be >0 except when the buffers are so long that it overflows the 32 bit output and produces a -ve result; which is obviously an error case.